typescript/sdk/src/ism/EvmIsmModule.ts

```diff
@@ -1,89 +1,727 @@
 
-import { EvmIsmCreator } from './EvmIsmCreator.js';
 import { EvmIsmReader } from './EvmIsmReader.js';
-import { IsmConfig } from './types.js';
+import {
+  AggregationIsmConfig,
+  DeployedIsm,
+  DeployedIsmType,
+  IsmConfig,
+  IsmType,
+  MultisigIsmConfig,
+  RoutingIsmConfig,
+  RoutingIsmDelta,
+} from './types.js';
+import { moduleMatchesConfig, routingModuleDelta } from './utils.js';
+
+type ExtraArgs = {
+  deployedIsm: Address;
+  mailbox: Address;
+};
 
 export class EvmIsmModule extends HyperlaneModule<
   ProtocolType.Ethereum,
   IsmConfig,
-  HyperlaneContracts<ProxyFactoryFactories> & {
-    deployedIsm: Address;
-  }
+  HyperlaneAddresses<ProxyFactoryFactories> & ExtraArgs
 > {
-  protected logger = rootLogger.child({ module: 'EvmIsmModule' });
-  protected reader: EvmIsmReader;
-  protected creator: EvmIsmCreator;
+  protected readonly logger = rootLogger.child({ module: 'EvmIsmModule' });
+  protected readonly reader: EvmIsmReader;
+  protected readonly factories: HyperlaneContracts<ProxyFactoryFactories>;
+
+  // Adding these to reduce how often we need to grab from MultiProvider.
+  public readonly chainName: string;
+  // We use domainId here because MultiProvider.getDomainId() will always
+  // return a number, and EVM the domainId and chainId are the same.
+  public readonly domainId: number;
 
   protected constructor(
     protected readonly multiProvider: MultiProvider,
     protected readonly deployer: HyperlaneDeployer<any, any>,
     args: HyperlaneModuleArgs<
       IsmConfig,
-      HyperlaneContracts<ProxyFactoryFactories> & {
-        deployedIsm: Address;
-      }
+      HyperlaneAddresses<ProxyFactoryFactories> & ExtraArgs
     >,
   ) {
     super(args);
+
     this.reader = new EvmIsmReader(multiProvider, args.chain);
-    this.creator = new EvmIsmCreator(deployer, multiProvider, args.addresses);
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { mailbox, deployedIsm, ...addresses } = args.addresses;
+    this.factories = attachAndConnectContracts(
+      addresses,
+      proxyFactoryFactories,
+      multiProvider.getSigner(args.chain),
+    );
+
+    this.chainName = this.multiProvider.getChainName(this.args.chain);
+    this.domainId = this.multiProvider.getDomainId(this.chainName);
   }
 
   public async read(): Promise<IsmConfig> {
-    return await this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
+    return this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
   }
 
-  public async update(config: IsmConfig): Promise<EthersV5Transaction[]> {
-    throw new Error('Method not implemented.');
+  // whoever calls update() needs to ensure that targetConfig has a valid owner
+  public async update(
+    targetConfig: IsmConfig,
+  ): Promise<AnnotatedEthersV5Transaction[]> {
+    const currentConfig =
+      typeof this.args.config === 'string'
+        ? this.args.addresses.deployedIsm
+        : await this.read();
```

Comment:
imo `this.read` should handle these branches

Reply:
good catch